cl/citations/match_citations.py

```diff
@@ -139,3 +147,120 @@ def match_citation(citation, citing_doc=None):
 
     # Give up.
     return []
+
+
+def get_citation_matches(citing_opinion, citations):
+    """For a list of Citation objects (e.g., FullCitations, SupraCitations,
+    IdCitations, etc.), try to match them to Opinion objects in in the database
+    using a variety of heuristics.
+
+    Returns:
+      - a list of Opinion objects, as matched to citations
+    """
+    citation_matches = []  # List of matches to return
+    was_matched = False  # Whether the previous citation match was successful
+
+    for citation in citations:
+        matched_opinion = None
+
+        # If the citation is to a non-opinion document, we currently cannot
+        # match these.
+        if isinstance(citation, NonopinionCitation):
+            pass
+
+        # If the citation is an id citation, just resolve it to the opinion
+        # that was matched immediately prior (so long as the previous match
+        # was successful).
+        elif isinstance(citation, IdCitation):
+            if was_matched:
+                matched_opinion = citation_matches[-1]
+
+        # If the citation is a supra citation, try to resolve it to one of
+        # the citations that has already been matched
+        elif isinstance(citation, SupraCitation):
+            candidates = []
+            for cm in citation_matches:
+                # The only clue we have to help us with resolution is the guess
+                # of what the supra citation's antecedent is, so we try to
+                # match that string to one of the known case names of the
+                # already matched opinions. However, because case names might
+                # look alike, matches using this heuristic may not be unique.
+                # If no match, or more than one match, is found, then the supra
+                # reference is effectively dropped.
+                if strip_punct(citation.antecedent_guess) in best_case_name(
```

Comment:
Seems like this punctuation stripping could be moved to the top? In a longer document, we're going to be running dozens of times. It'd also simplify the code, I think.

Reply:
Do you mean the top of the master `for` loop, i.e., around line 165? I can't put it there, unfortunately, because not all subclasses of `Citation` contain an `antecedent_guess` to be stripped. (So it has to go within each case of the appropriate `if`/`elif` blocks, at least. Though that's only twice.) I could separate the result of the stripping into its own variable, though, so that it isn't happening at the same time we're comparing it to the `best_case_name`?

Reply:
That seems good, yeah.

Reply:
Fixed in a67cef7